### ISSUE TITLE
homebrew_bin_path in homebrew_user doesn't deal with non-strings

### ISSUE THREAD

**Issue by cisco-abrandel** [chef/chef#13934]
## Description

After updating to chef-client 18.3.0 on MacOS, we ran into issues where chef-client was no longer converging due to a bug introduced in the homebrew_user helper. 

https://github.com/chef/chef/blob/v18.3.0/lib/chef/mixin/homebrew_user.rb#L64 calls which, which can return boolean values (specifically false here) as per https://github.com/chef/chef/blob/v18.3.0/chef-utils/lib/chef-utils/dsl/which.rb#L53

I think that the fix is that the code here needs to deal with all the possible return values of which, which includes booleans. 

## Chef Version

18.3.0

## Platform Version

MacOS 10.15

## Replication Case

Homebrew executable is located somewhere that is not in the PATH for the launchd that is running chef. 

## Client Output

```
[2023-09-19T16:47:45+00:00] ERROR: Running exception handlers
  - ChefRunRecorder::LastRunRecorder
Running handlers complete
[2023-09-19T16:47:45+00:00] ERROR: Exception handlers complete
Infra Phase failed. 10 resources updated in 05 minutes 18 seconds
[2023-09-19T16:47:45+00:00] FATAL: Stacktrace dumped to /var/chef/cache/chef-stacktrace.out
[2023-09-19T16:47:45+00:00] FATAL: ---------------------------------------------------------------------------------------
[2023-09-19T16:47:45+00:00] FATAL: PLEASE PROVIDE THE CONTENTS OF THE stacktrace.out FILE (above) IF YOU FILE A BUG REPORT
[2023-09-19T16:47:45+00:00] FATAL: ---------------------------------------------------------------------------------------
[2023-09-19T16:47:45+00:00] FATAL: TypeError: homebrew_package[openssl@1.1, autoconf, automake, libtool] (systems::packages line 228) had an error: TypeError: no implicit conversion of false into String
```

## Stacktrace

```
TypeError: homebrew_package[openssl@1.1, autoconf, automake, libtool] (systems::packages line 228) had an error: TypeError: no implicit conversion of false into String
/opt/chef/embedded/lib/ruby/gems/3.1.0/gems/chef-18.3.0/lib/chef/mixin/homebrew_user.rb:64:in `exist?'
/opt/chef/embedded/lib/ruby/gems/3.1.0/gems/chef-18.3.0/lib/chef/mixin/homebrew_user.rb:64:in `block in homebrew_bin_path'
/opt/chef/embedded/lib/ruby/gems/3.1.0/gems/chef-18.3.0/lib/chef/mixin/homebrew_user.rb:64:in `select'
/opt/chef/embedded/lib/ruby/gems/3.1.0/gems/chef-18.3.0/lib/chef/mixin/homebrew_user.rb:64:in `homebrew_bin_path'
/opt/chef/embedded/lib/ruby/gems/3.1.0/gems/chef-18.3.0/lib/chef/mixin/homebrew_user.rb:72:in `calculate_owner'
/opt/chef/embedded/lib/ruby/gems/3.1.0/gems/chef-18.3.0/lib/chef/mixin/homebrew_user.rb:47:in `find_homebrew_uid'
/opt/chef/embedded/lib/ruby/gems/3.1.0/gems/chef-18.3.0/lib/chef/provider/package/homebrew.rb:182:in `brew_cmd_output'
/opt/chef/embedded/lib/ruby/gems/3.1.0/gems/chef-18.3.0/lib/chef/provider/package/homebrew.rb:98:in `brew_info'
/opt/chef/embedded/lib/ruby/gems/3.1.0/gems/chef-18.3.0/lib/chef/provider/package/homebrew.rb:126:in `package_info'
/opt/chef/embedded/lib/ruby/gems/3.1.0/gems/chef-18.3.0/lib/chef/provider/package/homebrew.rb:146:in `installed_version'
/opt/chef/embedded/lib/ruby/gems/3.1.0/gems/chef-18.3.0/lib/chef/provider/package/homebrew.rb:52:in `block in get_current_versions'
/opt/chef/embedded/lib/ruby/gems/3.1.0/gems/chef-18.3.0/lib/chef/provider/package/homebrew.rb:51:in `map'
/opt/chef/embedded/lib/ruby/gems/3.1.0/gems/chef-18.3.0/lib/chef/provider/package/homebrew.rb:51:in `get_current_versions'
/opt/chef/embedded/lib/ruby/gems/3.1.0/gems/chef-18.3.0/lib/chef/provider/package/homebrew.rb:38:in `load_current_resource'
/opt/chef/embedded/lib/ruby/gems/3.1.0/gems/chef-18.3.0/lib/chef/provider.rb:228:in `run_action'
/opt/chef/embedded/lib/ruby/gems/3.1.0/gems/chef-18.3.0/lib/chef/resource.rb:601:in `block in run_action'
/opt/chef/embedded/lib/ruby/gems/3.1.0/gems/chef-18.3.0/lib/chef/resource.rb:628:in `with_umask'
/opt/chef/embedded/lib/ruby/gems/3.1.0/gems/chef-18.3.0/lib/chef/resource.rb:600:in `run_action'
/opt/chef/embedded/lib/ruby/gems/3.1.0/gems/chef-18.3.0/lib/chef/runner.rb:74:in `run_action'
/opt/chef/embedded/lib/ruby/gems/3.1.0/gems/chef-18.3.0/lib/chef/runner.rb:108:in `block in run_all_actions'
/opt/chef/embedded/lib/ruby/gems/3.1.0/gems/chef-18.3.0/lib/chef/runner.rb:108:in `each'
/opt/chef/embedded/lib/ruby/gems/3.1.0/gems/chef-18.3.0/lib/chef/runner.rb:108:in `run_all_actions'
/opt/chef/embedded/lib/ruby/gems/3.1.0/gems/chef-18.3.0/lib/chef/runner.rb:132:in `block in converge'
/opt/chef/embedded/lib/ruby/gems/3.1.0/gems/chef-18.3.0/lib/chef/resource_collection/resource_list.rb:96:in `block in execute_each_resource'
/opt/chef/embedded/lib/ruby/gems/3.1.0/gems/chef-18.3.0/lib/chef/resource_collection/stepable_iterator.rb:114:in `call_iterator_block'
/opt/chef/embedded/lib/ruby/gems/3.1.0/gems/chef-18.3.0/lib/chef/resource_collection/stepable_iterator.rb:85:in `step'
/opt/chef/embedded/lib/ruby/gems/3.1.0/gems/chef-18.3.0/lib/chef/resource_collection/stepable_iterator.rb:103:in `iterate'
/opt/chef/embedded/lib/ruby/gems/3.1.0/gems/chef-18.3.0/lib/chef/resource_collection/stepable_iterator.rb:54:in `each_with_index'
/opt/chef/embedded/lib/ruby/gems/3.1.0/gems/chef-18.3.0/lib/chef/resource_collection/resource_list.rb:94:in `execute_each_resource'
/opt/chef/embedded/lib/ruby/3.1.0/forwardable.rb:238:in `execute_each_resource'
/opt/chef/embedded/lib/ruby/gems/3.1.0/gems/chef-18.3.0/lib/chef/runner.rb:130:in `converge'
/opt/chef/embedded/lib/ruby/gems/3.1.0/gems/chef-18.3.0/lib/chef/client.rb:869:in `block in converge'
/opt/chef/embedded/lib/ruby/gems/3.1.0/gems/chef-18.3.0/lib/chef/client.rb:864:in `catch'
/opt/chef/embedded/lib/ruby/gems/3.1.0/gems/chef-18.3.0/lib/chef/client.rb:864:in `converge'
/opt/chef/embedded/lib/ruby/gems/3.1.0/gems/chef-18.3.0/lib/chef/client.rb:888:in `converge_and_save'
/opt/chef/embedded/lib/ruby/gems/3.1.0/gems/chef-18.3.0/lib/chef/client.rb:298:in `run'
/opt/chef/embedded/lib/ruby/gems/3.1.0/gems/chef-18.3.0/lib/chef/application.rb:305:in `run_with_graceful_exit_option'
/opt/chef/embedded/lib/ruby/gems/3.1.0/gems/chef-18.3.0/lib/chef/application.rb:281:in `block in run_chef_client'
/opt/chef/embedded/lib/ruby/gems/3.1.0/gems/chef-18.3.0/lib/chef/local_mode.rb:42:in `with_server_connectivity'
/opt/chef/embedded/lib/ruby/gems/3.1.0/gems/chef-18.3.0/lib/chef/application.rb:264:in `run_chef_client'
/opt/chef/embedded/lib/ruby/gems/3.1.0/gems/chef-18.3.0/lib/chef/application/base.rb:354:in `run_application'
/opt/chef/embedded/lib/ruby/gems/3.1.0/gems/chef-18.3.0/lib/chef/application.rb:67:in `run'
/opt/chef/embedded/lib/ruby/gems/3.1.0/gems/chef-bin-18.3.0/bin/chef-client:25:in `<top (required)>'
/opt/chef/bin/chef-client:180:in `load'
/opt/chef/bin/chef-client:180:in `<main>
```

**Comment by tpowell-progress** [comment]
This will be addressed with the next 18 release. Problem occurs if `brew` is not in the path, I believe, so ensuring `brew` exists in the path first is a potential workaround.

**Comment by jazaval** [comment]
Will this fix be released soon? It seems like the 18.2.46 release now has this issue as well, and previous 18.2.x release (18.2.7) our CI was pinned at seems to not be available via omnibus currently.

**Comment by tpowell-progress** [comment]
This was released (and subsequently patched for a slight miss) as of 18.4.2.
